pkg/operator/controller/canary/http.go

```diff
@@ -0,0 +1,118 @@
+package canary
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	routev1 "github.com/openshift/api/route/v1"
+
+	"github.com/tcnksm/go-httpstat"
+)
+
+const (
+	echoServerPortAckHeader = "x-request-port"
+)
+
+// probeRouteEndpoint probes the given route's host
+// and returns an error when applicable.
+func probeRouteEndpoint(route *routev1.Route) error {
+	host := route.Spec.Host
+
+	if len(host) == 0 {
+		return fmt.Errorf("route.Spec.Host is empty, cannot test route")
+	}
+
+	// Create HTTP request
+	request, err := http.NewRequest("GET", "http://"+host, nil)
+	if err != nil {
+		return fmt.Errorf("error creating canary HTTP request: %v", err)
+	}
+
+	// Create HTTP result
+	// for request stats tracking.
+	result := &httpstat.Result{}
+
+	// Get request context
+	ctx := httpstat.WithHTTPStat(request.Context(), result)
+	request = request.WithContext(ctx)
+
+	// Send the HTTP request
+	timeout, _ := time.ParseDuration("10s")
+	client := &http.Client{Timeout: timeout}
+	response, err := client.Do(request)
+
+	if err != nil {
+		// Check if err is a DNS error
+		dnsErr := &net.DNSError{}
+		if errors.As(err, &dnsErr) {
+			// Handle DNS error
+			CanaryRouteDNSError.WithLabelValues(host, dnsErr.Server).Inc()
+			return fmt.Errorf("error sending canary HTTP request: DNS error: %v", err)
+		}
+		// Check if err is a timeout error
+		if os.IsTimeout(err) {
+			// Handle timeout error
+			return fmt.Errorf("error sending canary HTTP Request: Timeout: %v", err)
+		}
+		return fmt.Errorf("error sending canary HTTP request: %v", err)
+	}
+
+	// Close response body even if read fails
+	defer response.Body.Close()
+
+	// Read response body
+	bodyBytes, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return fmt.Errorf("error reading canary response body: %v", err)
+	}
+	body := string(bodyBytes)
+	t := time.Now()
+	// Mark request as finished
+	result.End(t)
+	totalTime := result.Total(t)
+
+	// Verify body contents
+	if len(body) == 0 {
+		return fmt.Errorf("expected canary response body to not be empty")
+	}
+
+	expectedBodyContents := "Hello OpenShift!"
+	if !strings.Contains(body, expectedBodyContents) {
+		return fmt.Errorf("expected canary request body to contain %q", expectedBodyContents)
```

Comment:
log the actual too - may help diagnose a failure/pattern.

Reply:
Decided against this in the event that the router's 503 page is the actual body (since that's so long).